File: src/lib/__tests__/guide-course.test.ts
import { describe, it, expect } from "vitest";
import {
  courseStep,
  COURSE_ACC_MAX,
  COURSE_MIN_SPEED_MPS,
  COURSE_STALE_S,
} from "../guide-course";

const base = {
  course: 90,
  courseAccuracy: 10,
  speed: 1.2,
  motion: "moving" as const,
  ageSeconds: 1,
};

describe("courseStep", () => {
  it("모든 게이트를 통과하면 유효", () => {
    expect(courseStep(base)).toEqual({ kind: "valid", course: 90 });
  });

  it("course 음수는 실패(모름이 아니다)", () => {
    expect(courseStep({ ...base, course: -1 }).kind).toBe("invalid");
  });

  it("courseAccuracy 음수는 실패", () => {
    expect(courseStep({ ...base, courseAccuracy: -1 }).kind).toBe("invalid");
  });

  it("NaN은 실패로 떨어진다(부정 비교 계약)", () => {
    expect(courseStep({ ...base, course: NaN }).kind).toBe("invalid");
    expect(courseStep({ ...base, courseAccuracy: NaN }).kind).toBe("invalid");
    expect(courseStep({ ...base, speed: NaN }).kind).toBe("unknown");
    expect(courseStep({ ...base, ageSeconds: NaN }).kind).toBe("unknown");
  });

  it("courseAccuracy가 버킷 반폭을 넘으면 모름 — 존재만 확인하면 120도도 통과한다", () => {
    expect(courseStep({ ...base, courseAccuracy: COURSE_ACC_MAX + 0.1 }).kind).toBe(
      "unknown",
    );
    expect(courseStep({ ...base, courseAccuracy: COURSE_ACC_MAX }).kind).toBe("valid");
    expect(courseStep({ ...base, courseAccuracy: 120 }).kind).toBe("unknown");
  });

  it("정지 상태는 모름", () => {
    expect(courseStep({ ...base, motion: "stopped" }).kind).toBe("unknown");
    expect(courseStep({ ...base, motion: "speedUnknown" }).kind).toBe("unknown");
  });

  it("속도 하한 미달은 모름", () => {
    expect(courseStep({ ...base, speed: COURSE_MIN_SPEED_MPS - 0.01 }).kind).toBe(
      "unknown",
    );
    expect(courseStep({ ...base, speed: COURSE_MIN_SPEED_MPS }).kind).toBe("valid");
  });

  it("워치독 만료는 모름", () => {
    expect(courseStep({ ...base, ageSeconds: COURSE_STALE_S + 0.1 }).kind).toBe(
      "unknown",
    );
    expect(courseStep({ ...base, ageSeconds: COURSE_STALE_S }).kind).toBe("valid");
  });

  it("실패가 모름보다 앞선다 — 무효값은 정지 상태에서도 실패로 남는다", () => {
    expect(courseStep({ ...base, course: -1, motion: "stopped" }).kind).toBe("invalid");
  });
});

File: android/kit/src/test/kotlin/space/dodoplanet/gildongmu/kit/GuideCourseTest.kt
package space.dodoplanet.gildongmu.kit

import kotlin.test.Test
import kotlin.test.assertEquals

/** 진행 방향 3-state 판정 — Kit `GuideCourseTests`·웹 `guide-course.test.ts`와 같은 케이스 열. */
class GuideCourseTest {
    private fun step(
        course: Double = 90.0,
        courseAccuracy: Double = 10.0,
        speed: Double = 1.2,
        motion: MotionState = MotionState.moving,
        ageSeconds: Double = 1.0,
    ) = courseStep(course, courseAccuracy, speed, motion, ageSeconds)

    private val valid90 = CourseState.Valid(90.0)

    @Test fun `모든 게이트를 통과하면 유효`() {
        assertEquals(valid90, step())
    }

    @Test fun `무효값은 실패다 — 모름이 아니다`() {
        assertEquals(CourseState.Invalid, step(course = -1.0))
        assertEquals(CourseState.Invalid, step(courseAccuracy = -1.0))
    }

    @Test fun `NaN은 부정 비교로 걸러진다`() {
        assertEquals(CourseState.Invalid, step(course = Double.NaN))
        assertEquals(CourseState.Invalid, step(courseAccuracy = Double.NaN))
        assertEquals(CourseState.Unknown, step(speed = Double.NaN))
        assertEquals(CourseState.Unknown, step(ageSeconds = Double.NaN))
    }

    /** ⚠ 존재만 확인하면 120°도 통과해 반대 방향을 말한다. */
    @Test fun `버킷 반폭을 넘는 정확도는 모름`() {
        assertEquals(valid90, step(courseAccuracy = courseAccuracyMaxDegrees))
        assertEquals(CourseState.Unknown, step(courseAccuracy = courseAccuracyMaxDegrees + 0.1))
        assertEquals(CourseState.Unknown, step(courseAccuracy = 120.0))
    }

    @Test fun `정지·속도미상은 모름`() {
        assertEquals(CourseState.Unknown, step(motion = MotionState.stopped))
        assertEquals(CourseState.Unknown, step(motion = MotionState.speedUnknown))
    }

    @Test fun `속도 하한 미달은 모름`() {
        assertEquals(valid90, step(speed = courseMinSpeedMps))
        assertEquals(CourseState.Unknown, step(speed = courseMinSpeedMps - 0.01))
    }

    @Test fun `워치독 만료는 모름`() {
        assertEquals(valid90, step(ageSeconds = courseStaleSeconds))
        assertEquals(CourseState.Unknown, step(ageSeconds = courseStaleSeconds + 0.1))
    }

    @Test fun `실패가 모름보다 앞선다`() {
        assertEquals(CourseState.Invalid, step(course = -1.0, motion = MotionState.stopped))
    }
}
